### ISSUE TITLE
Large transfers on RAW drives fail

### ISSUE THREAD

**Issue by atarijookie** [atarijookie/ce-atari#152]
This is related to #100...

As expected, RAW media on USB suffer from the same issue (or one of them):

```
In    tpSCSICmd $5522
  Handle $333c6
  CmdLen 10  Cmd $28:00:00:00:04:98:00:6D:50:00
  TransferLen 14327808  Buffer $4d264
  SenseBuffer $32d42  Timeout 24000
  Flags $00
-> CHECK CONDITION  SenseKey $03  ASC $00  ASCQ $00
```

```
02956303 00000000 (2017-01-24 04:19:40.783776)	handleAcsiCommand: 5f 28 00 00 00 71 f0 00 41 48 00 00 00 00 isIcd
02956303 00000000 (2017-01-24 04:19:40.783878)	Scsi::SCSI_ReadWrite10 - READ, sector: 000071f0, count: 4148
02956303 00000000 (2017-01-24 04:19:40.783966)
     LOG_ERROR occurred
     Time since beginning of ACSI command handling: 0
     Time between this and previous ACSI command  : 194
02956303 00000000 (2017-01-24 04:19:40.783966)	Scsi::readSectors -- tried to read more than BUFFER_SIZE_SECTORS at once, fail!
02956303 00000000 (2017-01-24 04:19:40.784060)	Scsi::SCSI_ReadWrite10 - failed, returning status 02
02956305 00000002 (2017-01-24 04:19:40.785245)	

02956862 00000000 (2017-01-24 04:19:41.342520)	handleAcsiCommand: 5f 2a 00 00 01 20 90 00 41 48 00 00 00 00 isIcd
02956862 00000000 (2017-01-24 04:19:41.342622)	Scsi::SCSI_ReadWrite10 - WRITE, sector: 00012090, count: 4148
02956862 00000000 (2017-01-24 04:19:41.342710)
     LOG_ERROR occurred
     Time since beginning of ACSI command handling: 0
     Time between this and previous ACSI command  : 337
02956862 00000000 (2017-01-24 04:19:41.342710)	Scsi::writeSectors -- tried to write more than BUFFER_SIZE_SECTORS at once, fail!
02956862 00000000 (2017-01-24 04:19:41.342804)	Scsi::SCSI_ReadWrite10 - failed, returning status 02
02956863 00000001 (2017-01-24 04:19:41.343989)	
```

The first part, which needs fixing is that it doesn't fit in the data buffer in scsi.cpp and datatrans.cpp - the current implementation has 
- in scsi.cpp - 1 MB buffer for read and write, 1 MB buffer for verification
- in datatrans.cpp - 1 MB buffer for read, 1 MB buffer for write

- retarded fix: statically grow buffers from 4 MB total to 4*32 MB (128 MB) - one transfer can be up to 64k sectors, that's 32 MB
- less retarded fix - grow the buffers depending on need (1 MB, 4 MB, 8 MB, 16 MB, 32 MB) - wouldn't affect devices connected to STs, but would eat up lots of RAM on Falcon/TT connected devices
- good fix - alter scsi.cpp and datatrans.cpp to work on 1 MB sub-blocks

Other needed to be done:
- in Hans, detect READ(10), WRITE(10), VERIFY(10), extend the timeout on those commands to be based on transfered MBs....
- the current transfer size of read / write is stored on 24 bits, that's 16 MB, so if we want to support more, we need to alter RPi vs Hans commands - extend it from 24b to 32b - is this needed?

**Comment by atarijookie** [comment]
@mikrosk @tin-nl - guys, can a Falcon or TT have more than 16 MB of ST RAM? I'm asking because:
- only ST RAM is used for DMA transfers, right?
- I need to do more development if I need to support >= 16 MB transfers, so if you say no, then it means less development ;)
- Uwe S. mentioned that through some SCSI emulation on Hatari the HDDRIVER can try to make up to 32 MB transfers - SCSI READ(10) and WRITE(10) commands have this 32 MB size as maximum, but if real Falcon and TT can't have more than 16 MB (14 MB?) of ST RAM, then screw that large support...

**Comment by mikrosk** [comment]
Nope. Not even 16, only 14. And don't forget that 14 MB is a theoretical limit. You have to have screen buffers somewhere, too. :)

(and yes, you can use DMA only for buffers in ST-RAM)

**Comment by miniupnp** [comment]
I think the "grow buffer" as needed can be easily done

**Comment by atarijookie** [comment]
> I think the "grow buffer" as needed can be easily done

Yes, but it's a waste of memory, so I'll rather do the sub-transfers in a loop....

**Comment by miniupnp** [comment]
and probably a waste of time too :) I mean the performance is better if making several transfers continiously during the ACSI / SCSI data flow

**Comment by atarijookie** [comment]
Sub transfers in ce_main_app were done in d6d0c6c , not tested - will test during next free night.

Added timeout to RPi transfers of Hans in 62dbc1c , not tested. Also just timer used for timeout, this limits us to 19 MB transfer timeout, which should be fine with maximum 14 MB ST RAM size.

In theory - this should be all that's needed to fix this issue. Have to test it though...

**Comment by atarijookie** [comment]
```
07483431 00000000 (2017-01-26 04:22:48.652558)	handleAcsiCommand: 3f 28 00 00 00 00 c8 00 00 08 00 00 00 00 isIcd
07483431 00000000 (2017-01-26 04:22:48.652661)	Scsi::readSectors() - startSectorNo: 200, sectorCount: 8 -- will do 1 loops
07483437 00000006 (2017-01-26 04:22:48.658092)
     LOG_ERROR occurred
     Time since beginning of ACSI command handling: 6
     Time between this and previous ACSI command  : 120
07483437 00000006 (2017-01-26 04:22:48.658092)	CConSpi::txRx - trying to TX/RX 10 more bytes then allowed! Fix this!
07483634 00000197 (2017-01-26 04:22:48.855924)	

07483635 00000001 (2017-01-26 04:22:48.856039)	handleAcsiCommand: 3f 28 00 00 00 04 98 00 6d 48 00 00 00 00 isIcd
07483635 00000000 (2017-01-26 04:22:48.856143)	Scsi::readSectors() - startSectorNo: 1176, sectorCount: 27976 -- will do 14 loops
07492327 00008692 (2017-01-26 04:22:57.548093)
     LOG_ERROR occurred
     Time since beginning of ACSI command handling: 8693
     Time between this and previous ACSI command  : 203
07492327 00008692 (2017-01-26 04:22:57.548093)	waitForATN 03 fail - timeout
07493356 00001029 (2017-01-26 04:22:58.577094)
     LOG_ERROR occurred
     Time since beginning of ACSI command handling: 9722
     Time between this and previous ACSI command  : 203
07493356 00001029 (2017-01-26 04:22:58.577094)	waitForATN 03 fail - timeout
07494384 00001028 (2017-01-26 04:22:59.605094)
     LOG_ERROR occurred
     Time since beginning of ACSI command handling: 10750
     Time between this and previous ACSI command  : 203
07494384 00001028 (2017-01-26 04:22:59.605094)	waitForATN 03 fail - timeout
07495412 00001028 (2017-01-26 04:23:00.633092)
```

**Comment by atarijookie** [comment]
If the current version fails during read, it will still tell ST that it succeeded, because of sending read status on start (was fine with small blocks, doesn't make sense with large blocks). 
- alter sendData_start() to not set status
- Hans should ignore this status and wait for proper status on the end, like when doing write - add different STATE_DATA_READ command for long transfers?
- after long read - Hans will send ATN_GET_STATUS, RPi will reply with CMD_SEND_STATUS
- should not do dataTrans->sendDataAndStatus() at end of Scsi::processCommand() in every case, because R/W already sent the data
- more logging to SCSI processing, so it would be easier to spot - command, id, lun
- more logging inside Scsi::readSectors(), so could be easier to debug

**Comment by atarijookie** [comment]
More work done in 215ce1f24938b0cdb4c256405940d01a04b2d442 and 100342bbf032ce1264618bc8e56263da57cf2ebe , needs testing with HW...

**Comment by atarijookie** [comment]
More fixes done to the work in progress (I accidentally broke booting)...

The transfer from RAW drive stops around 3 MB of data, I have to check the signals with the logic analyzer to make sure it's not in Xilinx...

```
handleAcsiCommand: 3f 28 00 00 00 00 00 00 50 00 00 00 00 00 isIcd
Scsi::processCommand -- isIcd: 1, ACSI ID: 1, LUN: 0, CMD: 0x28 - READ(10)
Scsi::readSectors() - startSectorNo: 0x0, sectorCount: 0x5000 -- will do 11 loops
Scsi::readSectors() - will read sectorCountNow: 0x800, sectors to go: 0x4800
Scsi::readSectors() - will read sectorCountNow: 0x800, sectors to go: 0x4000
Scsi::readSectors() - will read sectorCountNow: 0x800, sectors to go: 0x3800
     LOG_ERROR occurred
     Time since beginning of ACSI command handling: 2561
     Time between this and previous ACSI command  : 11037819
waitForATN 03 fail - timeout
Scsi::readSectors() - dataTrans->sendData_transferBlock() failed for startSectorNo: 0x1800, sectorCountNow: 0x800
Scsi::processCommand -- data was already sent to Hans, just send status (0x02) to Hans
     LOG_ERROR occurred
     Time since beginning of ACSI command handling: 3561
     Time between this and previous ACSI command  : 11037819
waitForATN 05 fail - timeout
>>> Suspicious core cycle load -- cycle time: 3800 ms, load:  93 %
FW: Hans,  2016-12-29, LED is: 255, XI: 0x22
FW: Hans,  2016-12-29, LED is: 255, XI: 0x22
```

**Comment by atarijookie** [comment]
Transfers happen for the first 2 seconds, then the bus keeps waiting for something, then the bus signals get reset to idle after 7 or 8 seconds.

The bus reset is caused by main app failing the large transfer, doing Hans reset, while Hans still waits for data...

When the issue happens, the ACK stays low for a long period (until reset).
ACK usually is 400 - 500 ns.
REQ is 90 ns long (guessed from logic analyzer), or around 70 ns (read from scope).

> SCSI-II spec says:
> The target shall assert the REQ signal for a minimum of an assertion period.
> Assertion period - 90 ns

![ce_scsi_long_ack2](https://cloud.githubusercontent.com/assets/18427437/22452740/b7da7e3c-e779-11e6-8493-5c89968b8d76.png)

![ce_scsi_req](https://cloud.githubusercontent.com/assets/18427437/22453685/d82f4ff8-e780-11e6-9fd8-1bdbd8010e78.png)

**Comment by atarijookie** [comment]
Extended REQ to 240 ns, the issue still happens. It's hard to catch on logic analyzer...
- add software trigger on this issue
- log (in Hans) after how many bytes this happens 
- try slower bridge functions for this to see if anything changes
- try it with normal SCSI HDD

**Comment by atarijookie** [comment]
Alright, years later, with exxos finding this github reported issue, having some issues with scsi drives, decided to ping me. He has scsi drive, I have modified the test to work with any scsi device (not just CosmosEx). 
From multiple user tests with HDDRIVER and large files copying, they figured out that 7.99MB is the maximum transfer size that works on Atari Falcon. 
From the AHDI sources we see this:
`Put code back in because DMAC has been modified
; to have a 14-bit counter.`
14 it counter means that the maximum transfer size is 0x3fff (16383) sectors. That's one sector less than 8 MB.
So the maximum transfer size the CosmosEx must work with is 8 MB. HDD drivers on Atari, which access RAW media on CE device, must limit transfers on Falcon to 7.99 MB. 
Additionally, the SCSI code of CE_DD.PRG had fixed timeout of 1 or 3 seconds, while larger transfers need more time. This was modified to be configurable in my current branch.
